src/coreclr/jit/emit.cpp

```diff
@@ -8163,35 +8163,123 @@ CORINFO_FIELD_HANDLE emitter::emitSimd16Const(simd16_t constValue)
     return emitComp->eeFindJitDataOffs(cnum);
 }
 
-#if defined(TARGET_XARCH)
-CORINFO_FIELD_HANDLE emitter::emitSimd32Const(simd32_t constValue)
+#ifdef TARGET_XARCH
+//------------------------------------------------------------------------
+// emitSimdConst: Create a simd data section constant.
+//
+// Arguments:
+//    constValue - constant value
+//    attr       - The EA_SIZE for the constant type
+//
+// Return Value:
+//    A field handle representing the data offset to access the constant.
+//
+// Note:
+// Access to inline data is 'abstracted' by a special type of static member
+// (produced by eeFindJitDataOffs) which the emitter recognizes as being a reference
+// to constant data, not a real static field.
+//
+CORINFO_FIELD_HANDLE emitter::emitSimdConst(simd_t* constValue, emitAttr attr)
 {
-    unsigned cnsSize  = 32;
-    unsigned cnsAlign = cnsSize;
+    unsigned  cnsSize  = EA_SIZE(attr);
+    unsigned  cnsAlign = cnsSize;
+    var_types dataType = (cnsSize >= 8) ? emitComp->getSIMDTypeForSize(cnsSize) : TYP_FLOAT;
 
+#ifdef TARGET_XARCH
     if (emitComp->compCodeOpt() == Compiler::SMALL_CODE)
     {
         cnsAlign = dataSection::MIN_DATA_ALIGN;
     }
+#endif // TARGET_XARCH
 
-    UNATIVE_OFFSET cnum = emitDataConst(&constValue, cnsSize, cnsAlign, TYP_SIMD32);
+    UNATIVE_OFFSET cnum = emitDataConst(constValue, cnsSize, cnsAlign, dataType);
     return emitComp->eeFindJitDataOffs(cnum);
 }
 
-CORINFO_FIELD_HANDLE emitter::emitSimd64Const(simd64_t constValue)
+//------------------------------------------------------------------------
+// emitSimdConstCompressedLoad: Create a simd data section constant,
+//   compressing it if possible, and emit an appropiate instruction
+//   to load or broadcast the constant to a register.
+//
+// Arguments:
+//    constValue - constant value
+//    attr       - The EA_SIZE for the constant type
+//    targetReg  - The target register
+//
+void emitter::emitSimdConstCompressedLoad(simd_t* constValue, emitAttr attr, regNumber targetReg)
 {
-    unsigned cnsSize  = 64;
-    unsigned cnsAlign = cnsSize;
+    assert(EA_SIZE(attr) >= 8);
+    unsigned    cnsSize = EA_SIZE(attr);
+    instruction ins     = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
 
-    if (emitComp->compCodeOpt() == Compiler::SMALL_CODE)
+    for (unsigned size = 4; size < cnsSize; size *= 2)
     {
-        cnsAlign = dataSection::MIN_DATA_ALIGN;
+        // If the constant has zero upper elements/lanes, we can use a smaller load,
+        // because all scalar and vector loads from memory zero the upper.
+
+        simd_t val = {};
+        memcpy(&val, constValue, size);
```

Comment:
Why are we copying here, rather than comparing against `constValue + offset` directly?

Reply:
We're comparing the zeroes in the newly-initialized copy as well, up to the length of the input constant. It's specifically the zero-fill that matters here.

Reply:
But why do we need to check zeros that weren't there previously? That just seems like unnecessary work

Reply:
We're basically emulating what the smaller load would do. e.g. `movss` loads 4 bytes from memory and then zeroes the upper elements+lanes.  I start with a zero vector and write the scalar, then see if it's still a match for the constant we need to load.

Reply:
This seems unnecessarily complex and error prone. The simple `if` chain is much clearer and understandable as to what's going on and intended to happen, IMO.